core/src/main/java/org/jruby/runtime/invokedynamic/MathLinker.java

```diff
@@ -319,63 +323,40 @@ public static IRubyObject fixnum_op_minus_two(ThreadContext context, IRubyObject
     }
     
     public static IRubyObject floatOperator(ThreadContext context, IRubyObject caller, IRubyObject self, JRubyCallSite site, double value) throws Throwable {
-        String operator = site.name();
+        String operator = site.name;
         String opMethod = MethodIndex.getFastFloatOpsMethod(operator);
         String name = "float_" + opMethod;
-        MethodType type = methodType(IRubyObject.class, ThreadContext.class, IRubyObject.class, IRubyObject.class);
         MethodHandle target = null;
-        
-        if (target == null) {
-            type = type.insertParameterTypes(3, double.class);
-            target = lookup().findStatic(MathLinker.class, name, type);
-            target = insertArguments(target, 3, value);
-        }
+
+        if (target == null) target = findTargetImpl(name, IRubyObject.class, value);
 
         MethodHandle fallback = FLOAT_OPERATOR_FAIL;
-        fallback = insertArguments(fallback, 3, site, context.runtime.newFloat(value));
-        
-        MethodHandle test = FLOAT_TEST;
-        test = permuteArguments(test, methodType(boolean.class, ThreadContext.class, IRubyObject.class, IRubyObject.class), new int[] {2});
+        fallback = insertArguments(fallback, 3, site, value);
 
-        if (LOG_BINDING) LOG.debug(name + "\tFloat operation at site #" + site.siteID() + " (" + site.file() + ":" + site.line() + ") bound directly");
+        RubyClass classFloat = context.runtime.getFloat();
+        int gen = classFloat.getGeneration();
+        CacheEntry entry = searchWithCache(operator, caller, self.getMetaClass(), site);
+        if (entry == null || !entry.method.isBuiltin()) gen--; // invalid - fallback on slow-path
 
-        site.setTarget(guardWithTest(test, target, fallback));
+        MethodHandle test = FLOAT_TEST_WITH_GEN; // NOTE: since we invalidate on Float change FLOAT_TEST should be enough?!
+        test = insertArguments(test, 1, classFloat, gen);
+        test = permuteArguments(test, methodType(boolean.class, ThreadContext.class, IRubyObject.class, IRubyObject.class), ARG_2_TO_0);
+        if (LOG_BINDING) LOG.debug(name + "\tFloat operation at site #" + site.siteID + " (" + site.file() + ":" + site.line() + ") bound directly");
         
         // confirm it's a Float
         target = guardWithTest(test, target, fallback);
-        
-        // check Float reopening
-        target = ((SwitchPoint)context.runtime.getFloatInvalidator().getData())
-                .guardWithTest(target, fallback);
-        
+        target = ((SwitchPoint) classFloat.getInvalidator().getData()).guardWithTest(target, fallback);
```

Comment:
This is a bit confusing to me...why do we need to check both the SwitchPoint and the generation? They should both invalidate on class hierarchy changes.

Reply:
okay, will remove. 
this is smt I wasn't sure about due the comment history where it was noted that the callback invalidation approach wasn't working in the past well. so I did the generation check as well as an attempt to invalidate methods "early" (besides the class-level method invalidator)